internal/handlers/metrics.go

```diff
@@ -0,0 +1,144 @@
+	if err != nil {
+		return util.AppendUpMetric("", metrics.PodName, metrics.Namespace, 0),
+			fmt.Errorf("error creating request for %s: %w", url, err)
+	}
+	// Perform the HTTP request with context
+	resp, err := h.client.Do(req)
+	if err != nil {
+		// Return up=0 for failed request
+		return util.AppendUpMetric("", metrics.PodName, metrics.Namespace, 0),
+			fmt.Errorf("error scraping %s: %w", url, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		// Return up=0 for non-200 status code
+		return util.AppendUpMetric("", metrics.PodName, metrics.Namespace, 0),
+			fmt.Errorf("failed to scrape %s, status code: %d", url, resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		// Return up=0 for read errors
+		return util.AppendUpMetric("", metrics.PodName, metrics.Namespace, 0),
+			fmt.Errorf("error reading response from %s: %w", url, err)
+	}
+
+	// Append 'up=1' for successful scrape
+	labeledMetrics := util.AppendLabels(string(body), metrics.PodName, metrics.Namespace)
+
+	return util.AppendUpMetric(labeledMetrics, metrics.PodName, metrics.Namespace, 1), nil
+}
+
+// AggregateMetrics collects metrics from all pods concurrently and returns aggregated results.
+func (h *MetricsHandler) AggregateMetrics(ctx context.Context, pw *k8s.PodScrapeWatcher) ([]string, []string) {
+	var wg sync.WaitGroup
+	var respMu sync.Mutex
+	responses := []string{}
+	errors := []string{}
+	for podIP, metrics := range pw.GetPodMetricsEndpoints() {
+		wg.Add(1)
+
+		go func(podIP string, metrics k8s.PodScrapeDetails) {
+			defer wg.Done()
+
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				metricsResult, err := h.ScrapePodMetrics(ctx, podIP, metrics)
+				respMu.Lock()
+				defer respMu.Unlock()
+				if err != nil {
+					errors = append(errors, err.Error())
+				}
+				responses = append(responses, metricsResult)
+			}
+		}(podIP, metrics)
+	}
+
+	// Wait for all goroutines to complete.
+	wg.Wait()
+
+	return responses, errors
+}
+
+// ProxyMetrics aggregates metrics from all pods, appends pod metadata and 'up' metric, and returns them as text.
+func (h *MetricsHandler) ProxyMetrics(w http.ResponseWriter, r *http.Request, pw *k8s.PodScrapeWatcher) {
+	ctx := r.Context()
+	responses, errors := h.AggregateMetrics(ctx, pw)
+
+	w.Header().Set("Content-Type", "text/plain")
+
+	// If there are responses, write them to the response body.
+	if len(responses) > 0 {
+		// Join responses with a newline and write them.
+		writeResponse(w, strings.Join(responses, "\n"), http.StatusOK)
+		return
+	}
+
+	// If there are no responses, but there are errors, return the errors.
+	if len(errors) > 0 {
+		// Set the appropriate error status code.
+		writeResponse(w, strings.Join(errors, "\n"), http.StatusInternalServerError)
+		return
+	}
+
+	// If there are no responses and no errors, return an appropriate status.
+	w.WriteHeader(http.StatusNoContent)
```

Comment:
so this mainly if no pods are matching the watcher's filtering and you end up with an empty list of endpoints, returning 200 will get Prometheus to set `up=1`, no sure tho if this would be acceptable.